lib/msal-common/src/client/SilentFlowClient.ts

```diff
@@ -0,0 +1,161 @@
+            cacheRecord.idToken = this.fetchIdToken(homeAccountId, environment, cacheRecord.account.realm);
+            idTokenObj = new IdToken(cacheRecord.idToken.secret, this.config.cryptoInterface);
+            cacheRecord.accessToken = this.fetchAccessToken(homeAccountId, environment, requestScopes, cacheRecord.account.realm);
+            cacheRecord.refreshToken = this.fetchRefreshToken(homeAccountId, environment);
+
+            // If accessToken has expired, call refreshToken flow to fetch a new set of tokens
+            if (request.forceRefresh || (!!cacheRecord.accessToken && this.isTokenExpired(cacheRecord.accessToken.expiresOn))) {
+                // check if we have refreshToken
+                if (!!cacheRecord.refreshToken) {
+                    const refreshTokenClient = new RefreshTokenClient(this.config);
+                    const refreshTokenRequest: RefreshTokenRequest = {
+                        scopes: request.scopes,
+                        refreshToken: cacheRecord.refreshToken.secret,
+                        authority: request.authority
+                    };
+
+                    return refreshTokenClient.acquireToken(refreshTokenRequest);
+                }
+                // no refresh Token
+                else {
+                    throw ClientAuthError.createNoTokenInCacheError();
+                }
+            }
+
+            if (cacheRecord.accessToken === null) {
+                throw ClientAuthError.createNoTokenInCacheError();
+            }
+        }
+
+        // generate Authentication Result
+        return {
+            uniqueId: idTokenObj.claims.oid || idTokenObj.claims.sub,
+            tenantId: idTokenObj.claims.tid,
+            scopes: requestScopes.asArray(),
+            account: CacheHelper.toIAccount(cacheRecord.account),
+            idToken: cacheRecord.idToken.secret,
+            idTokenClaims: idTokenObj.claims,
+            accessToken: cacheRecord.accessToken.secret,
+            expiresOn: new Date(cacheRecord.accessToken.expiresOn),
+            extExpiresOn: new Date(cacheRecord.accessToken.extendedExpiresOn),
+            familyId: null,
+        };
+    }
+
+    /**
+     * fetches idToken from cache if present
+     * @param request
+     */
+    private fetchIdToken(homeAccountId: string, environment: string, inputRealm: string): IdTokenEntity {
+        const idTokenKey: string = CacheHelper.generateCredentialCacheKey(
+            homeAccountId,
+            environment,
+            CredentialType.ID_TOKEN,
+            this.config.authOptions.clientId,
+            inputRealm
+        );
+        return this.unifiedCacheManager.getCredential(idTokenKey) as IdTokenEntity;
+    }
+
+    /**
+     * fetches accessToken from cache if present
+     * @param request
+     * @param scopes
+     */
+    private fetchAccessToken(homeAccountId: string, environment: string, scopes: ScopeSet, inputRealm: string): AccessTokenEntity {
```

Comment:
This doesn't feel like the right name, as it implies that a fetch request is involved. What about `readAccessTokenFromCache`?

Reply:
Also, you should mark this as potentially returning `null`.

Reply:
We should turn on `strictNullChecks` ASAP.

Reply:
Will turn on strict null checks after next beta release.

Reply:
renamed function.